### DISCUSSION TITLE
Handling Provider Disposal of an AsyncNotifierProvider?

### DISCUSSION THREAD

**Discussion by PhilipHuang2** [rrousselGit/riverpod#3568] [Q&A] [answered]
So I started working with Riverpod and I want a AsyncNotifierProvider to call my server and return a set of reviews to work with.

The thing is that I don't use the provider in all my pages and from what I read, if I don't read my provider in a page, it will dispose of itself.  I also read that provider have a dispose method which can handle what happens to a provider when it is meant to be disposed.

The Problem is that the documentation on AsyncNotifierProvider using build_runner is sparse so I'm not sure where to start.

I tried to overwrite a dipose method but I can't find it, and I am not sure how to enable a dipose catcher in build runner.
Here is the relevant part of my current code.
```
@riverpod
class AsyncReview extends _$AsyncReview {
  @override
  FutureOr<AsyncReviewClass> build(String userID) async {
    debugPrint('Async Review Provider Initialized');
    ApiResult<Map<String, dynamic>> value =
        await PalatteApi().getReviewList(userID);
    // debugPrint(const JsonEncoder.withIndent('  ').convert(value.data));
    List<Review> reviews = [];
    value.data?["Data"].forEach((review) {
      Review newReview = Review.fromJson(review);
      reviews.add(newReview);
    });
    reviews.sort();
    AsyncReviewClass container = AsyncReviewClass(reviews: reviews);

    return container;
  }

  @override
  void _dipose() {
    debugPrint('Async Review Provider Disposed');
  }
```

Also please correct me if I am wrong about how riverpod works in terms of creation and diposal.

1. Enters a page that listens to the provider
2. The create method runs of the AsyncNotifierProvider
3. The page loads and does things
4.  leaving the page
5. The provider is marked for diposal
6. Whenever the compiler wants to, calls the dipose method, and resolves it, and deletes the provider.

**Comment by gdurandrexel** [comment, marked as answer]
You can find more info on this topic here: https://riverpod.dev/docs/concepts/provider_lifecycles

TLDR; in your build() method, use ref.onDispose().

**Comment by PhilipHuang2** [comment]
Oh now I feel bad that it was so easy to find.
Thank you so much.  For people searching the same problem here is my code below.
```
@riverpod
class AsyncReview extends _$AsyncReview {
  @override
  FutureOr<AsyncReviewClass> build(String userID) async {
    debugPrint('Async Review Provider Initialized');
    ApiResult<Map<String, dynamic>> value =
        await PalatteApi().getReviewList(userID);
    // debugPrint(const JsonEncoder.withIndent('  ').convert(value.data));
    List<Review> reviews = [];
    value.data?["Data"].forEach((review) {
      Review newReview = Review.fromJson(review);
      reviews.add(newReview);
    });
    reviews.sort();
    AsyncReviewClass container = AsyncReviewClass(reviews: reviews);

    ref.onDispose(_dipose);

    return container;
  }

  void _dipose() {
    debugPrint('Async Review Provider Disposed');
  }
  ```
 You give ref a callback through the onDipose function, and It does trigger when I leave the page watching the provider.
